code/go/internal/validator/semantic/validate_kibana_no_legacy_visualizations.go

```diff
@@ -0,0 +1,94 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package semantic
+
+import (
+	"path"
+
+	"github.com/elastic/kbncontent"
+	"github.com/elastic/package-spec/v2/code/go/internal/fspath"
+	"github.com/elastic/package-spec/v2/code/go/internal/pkgpath"
+	se "github.com/elastic/package-spec/v2/code/go/pkg/specerrors"
+)
+
+// ValidateKibanaNoLegacyVisualizations reports legacy Kibana visualizations in a package.
+func ValidateKibanaNoLegacyVisualizations(fsys fspath.FS) se.ValidationErrors {
+	var errs se.ValidationErrors
+
+	// Collect by-reference visualizations for reference later.
+	// Note: this does not include Lens, Maps, or Discover. That's okay for this rule because none of those are legacy
+	visFilePaths := path.Join("kibana", "visualization", "*.json")
+	visFiles, _ := pkgpath.Files(fsys, visFilePaths)
+
+	for _, file := range visFiles {
+		visJSON, err := file.Values("$")
+
+		if err != nil {
+			errs = append(errs, se.NewStructuredErrorf("error getting JSON: %w", err))
+			continue
+		}
+
+
+		visMap, ok := visJSON.(map[string]interface{})
+		if !ok {
+			errs = append(errs, se.NewStructuredErrorf("JSON of unexpected type %T in file %s", visJSON, file.Name()))
+			continue
+		}
+
+		desc, err := kbncontent.DescribeVisualizationSavedObject(visMap)
+		if err != nil {
+			errs = append(errs, se.NewStructuredErrorf("error describing visualization saved object: %w", err))
+			continue
+		}
+
+		if desc.IsLegacy() {
+			var editor string
+			if result, err := desc.Editor(); err == nil {
+				editor = result
+			}
+			errs = append(errs, se.NewStructuredErrorf("file \"%s\" is invalid: found legacy visualization \"%s\" (%s, %s)", fsys.Path(file.Path()), desc.Title(), desc.SemanticType(), editor))
+		}
+	}
+
+	dashboardFilePaths := path.Join("kibana", "dashboard", "*.json")
+	dashboardFiles, err := pkgpath.Files(fsys, dashboardFilePaths)
+	if err != nil {
+		errs = append(errs, se.NewStructuredErrorf("error finding Kibana dashboard files: %w", err))
+		return errs
+	}
+
+	for _, file := range dashboardFiles {
+		dashboardJSON, err := file.Values("$")
+		if err != nil {
+			errs = append(errs, se.NewStructuredErrorf("error getting dashboard JSON: %w", err))
```

Comment:
Should these and other errors report the file name, or is that provided by the framework?

Reply:
Ideally they should report the file name, as done in https://github.com/elastic/package-spec/pull/610/files#diff-eb7225937e41bbf5ac936b24e88d3ea808a74197d9eee117199b7a19052df16aR87. In practice I guess that it is unlikely to see these errors, so as you prefer, we can also polish this in a follow up.